key-share/src/valid.rs

```diff
@@ -0,0 +1,242 @@
+    /// `Err(err)` containing the error and the invalid value.
+    fn validate(self) -> Result<Valid<Self>, ValidateError<Self, Self::Error>>
+    where
+        Self: Sized,
+    {
+        Valid::validate(self)
+    }
+}
+
+/// Represents a type that can be constructed and validated from `Parts`
+///
+/// That can be particularly useful when vaidating `Parts` is cheaper than validating `Self`.
+///
+/// ## Example
+/// Suppose you have a struct `KeyShare` that consists of [`DirtyCoreKeyShare`](crate::DirtyCoreKeyShare) and some `AuxData`. In
+/// order to validate `KeyShare`, both core key share and aux data need to be validated separately and then they need to be
+/// checked for consistency. Now, if you already have `Valid<DirtyKeyShare>` and `Valid<AuxData>`, then you can skip their validation
+/// and only check that they're consistent.
+///
+/// ```rust
+/// use key_share::{Valid, Validate, ValidateFromParts};
+/// use generic_ec::Curve;
+///
+/// pub struct KeyShare<E: Curve> {
+///     core: key_share::DirtyCoreKeyShare<E>,
+///     aux: AuxData,
+/// }
+/// # pub struct AuxData { /* ... */ }
+/// # impl Validate for AuxData {
+/// #     type Error = std::convert::Infallible;
+/// #     fn is_valid(&self) -> Result<(), Self::Error> { Ok(()) }
+/// # }
+///
+/// # type InvalidKeyShare = Box<dyn std::error::Error>;
+/// // Validation for the whole key share can be expensive
+/// impl<E: Curve> Validate for KeyShare<E> {
+///     type Error = InvalidKeyShare;
+///     fn is_valid(&self) -> Result<(), Self::Error> {
+///         self.core.is_valid()?;
+///         self.aux.is_valid()?;
+///         check_consistency(&self.core, &self.aux)
+///     }
+/// }
+/// fn check_consistency<E: Curve>(
+///     core: &key_share::DirtyCoreKeyShare<E>,
+///     aux: &AuxData,
+/// ) -> Result<(), InvalidKeyShare> {
+///     // check that `core` and `aux` seem to match each other
+/// # Ok(())
+/// }
+///
+/// // Sometimes, we already validated that `core` and `aux` are valid, so we can perform cheaper validation:
+/// impl<E: Curve> ValidateFromParts<(Valid<key_share::DirtyCoreKeyShare<E>>, Valid<AuxData>)>
+///     for KeyShare<E>
+/// {
+///     fn validate_parts(parts: &(Valid<key_share::DirtyCoreKeyShare<E>>, Valid<AuxData>)) -> Result<(), Self::Error> {
+///         check_consistency(&parts.0, &parts.1)
+///     }
+///     fn from_parts(parts: (Valid<key_share::DirtyCoreKeyShare<E>>, Valid<AuxData>)) -> Self {
+///         Self { core: parts.0.into_inner(), aux: parts.1.into_inner() }
+///     }
+/// }
+/// ```
+pub trait ValidateFromParts<Parts>: Validate {
+    /// Validates parts
+    ///
+    /// Note: implementation **must** guarantee that if `T::validate_parts(parts).is_ok()` then `T::from_parts(parts).is_valid().is_ok()`
+    fn validate_parts(parts: &Parts) -> Result<(), Self::Error>;
+    /// Constructs `Self` from parts
+    fn from_parts(parts: Parts) -> Self;
+}
```

Comment:
Do we really need this trait if there's going to be only one implementor? I'd rather have a trait `FromParts` implemented for valid and dirty shares, one checking for validity and the other not. This starts to look suspiciously similar to what we had before with TryFrom... But I'm really _not_ suggesting to go back to that, I like when my go-to-definition actually works

Reply:
Ah, I forgot that AuxInfo is not part of keyshare anymore and can be anything, which makes a `from_parts` a nice convenient method on Valid. Although I don't care about it for KU23 for example, because core share is all I need.

Reply:
If you don't care, don't use it :D For cggmp21 it's needed